### PR TITLE
Abstract realm cache clear for role mappers

### PR DESCRIPTION
**Pull request by albertzaharovits**
This is a refactoring and a minor bug-fix.
This creates a base class (`AbstractRoleMapperClearRealmCache`) that
the concrete role mappers (currently only `DnRoleMapper` and `NativeRoleMappingStore`) extend.
The abstract base class is responsible for storing references to realms
that need to be "refreshed" when the role mapper's contents change, and
to expose methods to trigger the "refreshes".

### CODE REVIEW

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/NativeRoleMappingStore.java [hunk 1]
```diff
@@ -384,30 +382,10 @@ public void onSecurityIndexStateChange(SecurityIndexManager.State previousState,
...
-        }
-    }
-
-    private <Result> void refreshRealms(ActionListener<Result> listener, Result result) {
-        if (realmsToRefresh.isEmpty()) {
-            listener.onResponse(result);
-            return;
+            // the notification that the index state changed is received on every node
+            // this means that here we need only to invalidate the local realm caches only
+            clearRealmCachesOnLocalNode();
```
Comment (albertzaharovits) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/NativeRoleMappingStore.java:387]: This is almost a bug fix.

Comment (gwbrown) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/support/mapper/NativeRoleMappingStore.java:387]: Just double checking my understanding: Previously, this receive a state change notification, and then every node would send out a message telling all nodes to invalidate their caches. Now, every nodes gets the state change and invalidates their caches locally. Correct?

Assuming I read that right, damn, nice job.

**Review by AthenaEryma — APPROVED** [review]
LGTM! Left one nitpicky ask for Javadoc but I'm not even gonna block on it.

**Comment by albertzaharovits** [comment]
> Just double checking my understanding: Previously, this receive a state change notification, and then every node would send out a message telling all nodes to invalidate their caches. Now, every nodes gets the state change and invalidates their caches locally. Correct?
Assuming I read that right, damn, nice job.

Yes, that's correct! I'll be explaining this is a short code comment as well.

**Comment by albertzaharovits** [comment]
@gwbrown do you have any reservations on not backporting this to 8.14?
I'm tempted to backport it given the bugfix-ish nature of it.

**Comment by albertzaharovits** [comment]
@gwbrown I've also pushed a rename `refreshRealmOnChange` -> `clearRealmCacheOnChange` after you reviewed (in order to not make it tedious for you to review). Let me know if you have any other naming suggestion.

**Comment by gwbrown** [comment]
> do you have any reservations on not backporting this to 8.14?

Let's do it. Seems safe to me.

Rename LGTM too!

**Pull request merged**